JetNews/app/src/main/java/com/example/jetnews/ui/interests/InterestsScreen.kt

```diff
@@ -94,80 +96,114 @@ class TabContent(val section: Sections, val content: @Composable () -> Unit)
-    // Pass them to the stateless InterestsScreen using a tabContent.
-    val topicsSection = TabContent(Sections.Topics) {
-        val selectedTopics by interestsViewModel.selectedTopics.collectAsState()
-        TopicList(
-            topics = uiState.topics,
-            selectedTopics = selectedTopics,
-            onTopicSelect = { interestsViewModel.toggleTopicSelection(it) }
-        )
+    val tabContent = rememberTapContent(interestsViewModel)
+    val (currentSection, updateSection) = rememberSaveable {
+        mutableStateOf(tabContent.first().section)
     }
 
-    val peopleSection = TabContent(Sections.People) {
-        val selectedPeople by interestsViewModel.selectedPeople.collectAsState()
-        PeopleList(
-            people = uiState.people,
-            selectedPeople = selectedPeople,
-            onPersonSelect = { interestsViewModel.togglePersonSelected(it) }
-        )
-    }
-
-    val publicationSection = TabContent(Sections.Publications) {
-        val selectedPublications by interestsViewModel.selectedPublications.collectAsState()
-        PublicationList(
-            publications = uiState.publications,
-            selectedPublications = selectedPublications,
-            onPublicationSelect = { interestsViewModel.togglePublicationSelected(it) }
-        )
-    }
-
-    val tabContent = listOf(topicsSection, peopleSection, publicationSection)
-    val (currentSection, updateSection) = rememberSaveable { mutableStateOf(tabContent.first().section) }
-    InterestsScreen(
+    InterestsScreenAdaptive(
         tabContent = tabContent,
-        tab = currentSection,
+        currentSection = currentSection,
+        showNavRail = showNavRail,
         onTabChange = updateSection,
+        navigateToHome = navigateToHome,
         openDrawer = openDrawer,
         scaffoldState = scaffoldState
     )
 }
 
+/**
+ * Stateless interest screen displays the tabs specified in [tabContent] adapting the UI to
+ * different screen sizes.
+ *
+ * @param tabContent (slot) the tabs and their content to display on this screen, must be a
+ * non-empty list, tabs are displayed in the order specified by this list
+ * @param currentSection (state) the current tab to display, must be in [tabContent]
+ * @param showNavRail (state) whether the Drawer or NavigationRail needs to be shown
+ * @param onTabChange (event) request a change in [currentSection] to another tab from [tabContent]
+ * @param navigateToHome (event) request navigation to Home screen
+ * @param openDrawer (event) request opening the app drawer
+ * @param scaffoldState (state) the state for the screen's [Scaffold]
+ */
+@Composable
+private fun InterestsScreenAdaptive(
+    tabContent: List<TabContent>,
+    currentSection: Sections,
+    showNavRail: Boolean,
+    onTabChange: (Sections) -> Unit,
+    navigateToHome: () -> Unit,
+    openDrawer: () -> Unit,
+    scaffoldState: ScaffoldState
+) {
+    if (showNavRail) {
+        Row(Modifier.fillMaxSize()) {
```

Comment:
I don't know the implications of this, but is there a difference between putting this `if` as close as possible to the content that is removed? For example, I think we could always have a `Row`, and then only have one place where `InterestsScreen` is defined (probably with an `if` statement in the arguments for the `navigationIconContent`).

Is there an advantage one way or the other? Or just situational based on other requirements (animations, etc.)

Reply:
The point of having the if statement above `Row` is to avoid the `Row` composable when it's not needed (i.e. in Compact mode).

This might be a micro-optimization though, but I think this reads better than using always the `Row` and having to call the `InterestsScreen` like:

```
InterestsScreen(
    tabContent = tabContent,
    tab = currentSection,
    onTabChange = onTabChange,
    scaffoldState = scaffoldState,
    navigationIconContent = if (!showNavRail) { {
        IconButton(onClick = openDrawer) {
            Icon(
                painter = painterResource(R.drawable.ic_jetnews_logo),
                contentDescription = stringResource(
                    R.string.cd_open_navigation_drawer
                ),
                tint = MaterialTheme.colors.primary
            )
        }
    } } else {
        null
    }
)
```

WDYT?

Reply:
Makes sense right now, although this is an interesting quirk of Compose I don't think I was aware of until now:

Let's say that `InterestsScreen` had some internal state (animations, scroll state, etc.)

Having it defined twice in separate locations means that that state would be reset if the one that is displayed switches. Having them in the same place (defined once, with the `if` statements within the declaration) I believe means that internal state will be preserved.

Reply:
Very good point, I need to change this! Thank you